core/src/main/scala/org/apache/spark/broadcast/TorrentExecutorBroadcast.scala

```diff
@@ -0,0 +1,174 @@
+ * In contrast, this implementation performs broadcast on executor side for a RDD. So the results
+ * of the RDD does not need to collect first back to the driver before broadcasting.
+ *
+ * The mechanism is as follows:
+ *
+ * On each executor, the executor first attempts to fetch the object from its BlockManager. If
+ * it doesn not exist, it then uses remote fetches to fetch the blocks of the RDD from other
+ * executors if available. Once it gets the blocks, it puts the blocks in its own BlockManager,
+ * ready for other executors to fetch from.
+ *
+ * @tparam T The type of the element of RDD to be broadcasted.
+ * @tparam U The type of object transformed from the collection of elements of the RDD.
+ *
+ * @param numBlocks Total number of blocks this broadcast variable contains.
+ * @param rddId The id of the RDD to be broadcasted on executors.
+ * @param mode The [[BroadcastMode]] object used to transform the result of RDD to the object which
+ *             will be stored in the [[BlockManager]].
+ * @param id A unique identifier for the broadcast variable.
+ */
+private[spark] class TorrentExecutorBroadcast[T: ClassTag, U: ClassTag](
+    numBlocks: Int,
+    rddId: Int,
+    mode: BroadcastMode[T],
+    id: Long) extends Broadcast[U](id) with Logging with Serializable {
+
+  /**
+   * Value of the broadcast object on executors. This is reconstructed by [[readBroadcastBlock]],
+   * which builds this value by reading blocks from other executors.
+   */
+  @transient private lazy val _value: U = readBroadcastBlock()
+
+  private val broadcastId = BroadcastBlockId(id)
+
+  override protected def getValue() = {
+    _value
+  }
+
+  /** Fetch torrent blocks from other executors. */
+  private def readBlocks(): Array[T] = {
+    // Fetch chunks of data. Note that all these chunks are stored in the BlockManager and reported
+    // to the driver, so other executors can pull these chunks from this executor as well.
+    val blocks = new Array[Array[T]](numBlocks)
+    val bm = SparkEnv.get.blockManager
+
+    for (pid <- Random.shuffle(Seq.range(0, numBlocks))) {
+      val pieceId = RDDBlockId(rddId, pid)
+      // First try getLocalValues because there is a chance that previous attempts to fetch the
+      // broadcast blocks have already fetched some of the blocks. In that case, some blocks
+      // would be available locally (on this executor).
+      bm.getLocalValues(pieceId) match {
+        case Some(block: BlockResult) =>
+          blocks(pid) = block.data.asInstanceOf[Iterator[T]].toArray
+        case None =>
+          bm.get[T](pieceId) match {
+            case Some(b) =>
+              val data = b.data.asInstanceOf[Iterator[T]].toArray
+              // We found the block from remote executors' BlockManager, so put the block
+              // in this executor's BlockManager.
+              if (!bm.putIterator(pieceId, data.toIterator,
```

Comment:
For RDD, there is a cleaning mechanism that the persisted pieces will be removed once the RDD is not referred. Because we fetch and use RDD pieces here instead of broadcast pieces in driver side broadcast, I think it should be fine to deliver the cleaning to current mechanism.